hackathon_site/hardware/test_api.py

```diff
@@ -1258,3 +1258,70 @@ def test_no_remaining_quantities(self):
             ],
         }
         self.assertEqual(response.json(), expected_response)
+
+
+class OrderListPatchTestCase(SetupUserMixin, APITestCase):
+    def setUp(self):
+        super().setUp()
+        self.team = Team.objects.create()
+        self.view_name = "api:hardware:order-detail"
+        self.change_permissions = Permission.objects.filter(
+            content_type__app_label="hardware", codename="change_order"
+        )
+        hardware = Hardware.objects.create(
+            name="name",
+            model_number="model",
+            manufacturer="manufacturer",
+            datasheet="/datasheet/location/",
+            quantity_available=4,
+            max_per_team=1,
+            picture="/picture/location",
+        )
+        order = Order.objects.create(status="Submitted", team=self.team)
+        OrderItem.objects.create(order=order, hardware=hardware)
+        self.pk = order.id
+
+    def _build_view(self, pk):
+        return reverse(self.view_name, kwargs={"pk": pk})
+
+    def test_user_not_logged_in(self):
+        request_data = {"status": "Ready for Pickup"}
+        response = self.client.patch(self._build_view(self.pk), request_data)
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+
+    def test_user_lack_perms(self):
+        self._login()
+        request_data = {"status": "Ready for Pickup"}
+        response = self.client.patch(self._build_view(self.pk), request_data)
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_successful_status_change(self):
+        self._login(self.change_permissions)
+        request_data = {"status": "Ready for Pickup"}
+        response = self.client.patch(self._build_view(self.pk), request_data)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(request_data["status"], Order.objects.get(id=self.pk).status)
+
+    def test_unallowed_status_change(self):
+        self._login(self.change_permissions)
+        request_data = {"status": "Picked Up"}
+        response = self.client.patch(self._build_view(self.pk), request_data)
+        self.assertEqual(
+            response.json(),
+            {
+                "detail": [
+                    "Cannot change the current status of the order to the desired order."
+                ]
+            },
+        )
+        self.assertFalse(request_data["status"] == Order.objects.get(id=self.pk).status)
+
+    def test_failed_beginning_status(self):
+        self._login(self.change_permissions)
+        request_data = {"status": "Cancelled"}
+        order = Order.objects.create(status="Picked Up", team=self.team)
+        response = self.client.patch(self._build_view(order.id), request_data)
+        self.assertEqual(
+            response.json(), {"detail": ["Cannot change the status for this order."]},
```

Comment:
I think it would be better to probably use arrays to store these "detail sentences" at the beginning of the test case. If we would like to change the prompt info in the future, it would be much easier to do that. How do you think Arsh?

Reply:
Could you clarify what you mean by "prompt"?

Reply:
oops I used the wrong word. I am referring to those "detail sentences": "Cannot change the status for thsi order"

Reply:
Hmmm I don't think it would be helpful. A list array isn't really beneficial as it doesn't state which test requires which detail statement.

A dict would be a better choice that maps a test name string to a detail statement, but even then, it may be better visually to see the expected detail statement within the actual subtest.

Furthermore, we haven't really adopted that convention (the list array convention) in many other test cases, and I'm not sure if it's a good idea to cause that significant change here.

Reply:
yeah, that's right. Dict would be better because we could map the test name easier. Something like this:
details = 
{
   "cannot_change _order" : "Cannot change the status for this order",
}

Reply:
@grahamhoyes what do you think? I personally don't wanna change, but as it is a recommendation that differs from typically how we do things in the django tests, I thought your input would be useful here

Reply:
I agree with Arsh, I don't think it makes sense to store the them at the test class level and index into them for each test for two reasons:
1. Each detail string only appears once in the `OrderListPatchTestCase` anyway, so there's only one place to change them now anyway.
2. It's best to keep things defined as close to the test assertion as possible, since it makes things easier to follow (unless it's truly generic to all the methods, like `self.view_name` or something). In this case, if a test failed because that detail string changed, we can see directly from the test output what the issue is.